### ISSUE TITLE
Kubernetes names of ServiceClass and Plan versus OSB Guids

### ISSUE THREAD

**Issue by duglin** [kubernetes-incubator/service-catalog#802]
See: https://docs.cloudfoundry.org/services/managing-service-brokers.html

Upon refreshing a catalog from a broker a Service and/or Plan name can change. This is problematic for us because we map those names to Kube resource names, which are immutable.

I think we need to use the "id" from the Broker as the Kube "name" since in both cases they'll be immutable. We can then create a property called OSBName (or something like that) to hold the "name" returned from the broker.  In UIs we need to remind people to show the OSBName not the Kube name.

**Comment by pmorie** [comment]
I agree with what you've said here if we can pin down for certain whether the names of these entities can change.  If you just look at the OSB spec, it's not super clear: https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#catalog-management

The link says not to change the ID of services and plans, which I would agree implies that the names _might_ be changed.  I think we should clarify this in the spec before we make any code changes.

If we do this, it's not an amazing experience if you're just writing yaml, but porcelain commands and good user interface can make this easier.

**Comment by pmorie** [comment]
@vaikas-google your thoughts appreciated here

**Comment by duglin** [comment]
yep - I shutter at the idea of asking people to type GUIDs instead of nicer names in their kube yaml but we may not have a choice.

**Comment by pmorie** [comment]
Opened https://github.com/openservicebrokerapi/servicebroker/issues/187 to determine the answer to this

**Comment by arschles** [comment]
We discussed in the SIG call that, if brokers are allowed to change names on catalog refreshes, we need to use the GUIDs, and reference them in `Instance`s. This is of course a poor user experience, so I propose that we build a layer of abstraction (the solution to everything in CS, of course!) for naming. I've detailed it in #805

**Comment by MHBauer** [comment]
One of the original conversations was to use the osb-guid as the k8s-meta-name for exactly this reason. We ended up using the osb-name for the usability reason.

**Comment by duglin** [comment]
On today's call we agreed to using ID as the metadata.name on ServiceClass and ServicePlan but we should keep `ExternalID` too just in case we decide to use something else for the name.

I'll open a new issue to discuss possible designs for how to add on something for a better UX - and that discuss (IMO) should happen before beta.

**Comment by jwforres** [comment]
On Thu, May 11, 2017 at 5:02 PM, Doug Davis <notifications@github.com>
wrote:

> On today's call we agreed to using ID as the metadata.name on
> ServiceClass and ServicePlan but we should keep ExternalID too just in
> case we decide to use something else for the name.
>
> I'll open a new issue to discuss possible designs for how to add on
> something for a better UX - and that discuss (IMO) should happen before
> beta.
>

Agreed about that happening before beta, please include me in that

**Comment by pmorie** [comment]
It looks like this isn't going to be in 0.0.7, moving to 0.1.0

**Comment by arschles** [comment]
If we switch to using GUIDs as the primary key for serviceclasses and plans, I'd like to include a way to assign human-readable names to the services and plans in the same change. I agree that a porcelain CLI may solve the naming problem, but it won't be a complete solution because `Instance`s will be created from helm charts as well.

**Comment by duglin** [comment]
while the most common pattern in kube for referencing other objects is by name, what other mechanisms are there?  I know labelSelectors is one but are there others?

**Comment by pmorie** [comment]
The only way to reference individual resources from another resource is by name.

I would be very curious to know how CF deals with name changes for services / plans.  Does someone want to take an action item to find out about that?

**Comment by duglin** [comment]
they expose the GUIDs for things so even if the user says "foo", they'll use the guid under the covers.

**Comment by duglin** [comment]
this is why using "names" as keys is a bad idea.  Typically names are human readable and may want to be changed from time to time - but keys/guids/ids/.... being static for all time is usually ok.

**Comment by duglin** [comment]
Just wondering... are there any cases in kube of people being allowed to modify the name property on a resource?  While not common if its technically allowed it might be something to consider.

**Comment by pmorie** [comment]
> Just wondering... are there any cases in kube of people being allowed to modify the name property on a resource? While not common if its technically allowed it might be something to consider.

No

**Comment by duglin** [comment]
I think its fair to point out that kube does reference other resources by things other than names - I consider labelSelector an example of that.

**Comment by pmorie** [comment]
Label selectors are not a way to reference individual objects

**Comment by duglin** [comment]
Some rough notes/ideas I've been thinking about for this one:

Assumptions:
- can't use serviceID or planID as kube resource names
  - therefore need to generate a kube acceptable names (eg uuid) for svc/plan
- want user to use human friendly service name and plan name
  - will need a mapping from friendly svc/plan name to kube name (uuid)
- when broker changes svc or plan name, all existing instances should no longer show the old name

Proposal:
- as service is registered, we create a ServiceClass:
  - name = new UUID()
  - uxName = osbapi.svc.name
  - `kubectl get serviceclasses` will not work for us until the results are generated server-side
  - in the meantime, we need to get a kubectl plugin written:
      `kubectl plugin marketplace`
    this will allow us to show the uxNames for services and plans
  
- as a plan is registered, we create a ServicePlan:
  - name = new UUID()
  - uxName = osbapi.plan.name
  
- when a user creates an Instance they use the human friendly names:
  - instance.service = svc.uxName
  - instance.plan    = svc.plan.uxName
  
  during processing of POST we:
  - convert instance.service to svc.name (the service's uuid) using the mapping defined above
  - convert instance.plan to svc.plan.name (the plan's uuid) using the mapping defined above
  
- when a user retrieves an Instance:
  - option 1: (preferred, if kube allows this)
    - we dynamically set: instance.serviceName (not instance.service) to svc.uxName
    - we dynamically set: instance.planName (not instance.plan) to svc.plan.uxName
  - option 2:
    - do nothing and people will need to use the service or plan fields
      as search criteria to retrieve the Service or Plan to see the
      real name and other bits of data

**Comment by pmorie** [comment]
I suggest that we avoid thinking too much about the technical details of the approach we eventually take until we have some more information on https://github.com/openservicebrokerapi/servicebroker/issues/273.

We should agree on what problems we're trying to solve.  Some preliminary things to come to an agreement on:

1.  Determine the correct kubernetes names of service classes and plans:
    1.  OSB IDs?
    2.  A synthesized, stable name?
    3.  Something else?
2.  Ideally, the references users make to services and plans should have be using human-readable (or at least human-meaningful) coordinates
3.  Determine what the right level for (2) is:
    1.  Is porcelain sufficient to solve this from a UX standpoint?
    2.  Is a raw yaml (ie, API construction) solution needed?
4.  Once we know the answer to (3), we can think about what kind of technical solution is needed

This is not a topic that we can simply talk through in a single design meeting.  In k8s core, I would insist that we tackle this with the type of proposal we typically write for large features.  Specific order of operations I suggest:

1.  Come to an agreement on the first 3 questions raised above
2.  Get definitive information about the validations CF applies from https://github.com/openservicebrokerapi/servicebroker/issues/273
3.  A single person should be assigned to write a formal proposal (possibly to the community repo) to outline the problem space, use cases, analysis, technical solution, and implementation details
4.  One or more people implement the proposal from (3)

I would be in favor of treating this as a general problem and making the proposal in the community repo.  My generic formulation of the problem is:

> Given some external entity without a human-readable stable coordinate for which there is an API resource X, what is the correct way of providing a human-meaningful stable coordinate to use as the kubernetes name for that entity?

We are likely eventually to have the same problem in another area of the system.  Let's do the work to vet the pattern we use and give others a template to solve it the same way.

Additional notes:

Re: https://github.com/kubernetes-incubator/service-catalog/issues/802#issuecomment-317299121:

I think we should avoid the dynamic transformation in the API server route:

- we do not do the type of dynamic transformation of data anywhere else in the system, so this would be establishing a new precedent
- API precedent is extraordinarily expensive to establish
- the API machinery does not support the kind of plumbing (access to resource X in resource Y's registry implementation) that would be required to do this

**Comment by pmorie** [comment]
Here's another requirement that I think exists:

- It should be possible to write a template, helm chart, resource YAML that reliably addresses a particular service class and plan that will be usable in multiple clusters

**Comment by pmorie** [comment]
@smarterclayton curious to know your thoughts here

**Comment by pmorie** [comment]
One thought I had: it would be possible conceivably to support different modes of addressing a serviceclass and plan with a union type.

Also, I would like to point out that this is related to #1010

**Comment by duglin** [comment]
I think there are multiple ways in which to view how a service would be used across clusters.

1 - from within each cluster there's a cross-cluster reference to the service - @pmorie I think this is the mode you're referring to in your comment above. And in that case I would imagine the "normal" Kube mechanism would be used. Meaning, the reference would need to include some kind of cluster ID/name.  In this case though I would imagine that the service would need to be owned by just one cluster and the one would be responsible for managing it. 

2 - if people want to share services across clusters then the service is registered with each cluster independently. This has the advantage of the admin being able to managed which clusters sees which service/plan. This is the model we use in Bluemix today except its not just across Kube clusters but even CF and Docker clusters.

I'm not sure I see the advantage of the first model since there really isn't any logic associated with the registration process - its more a question of visibility and access control - which is better done via the 2nd model. The Broker is really the thing that's shared.

**Comment by pmorie** [comment]
I did not have cross-cluster in mind as a usage model at all, actually, might need to go and clarify some of my above comment

**Comment by pmorie** [comment]
Let me clarify my above comments.  So far in our discussion of this issue we have implicitly been hunting for a single 'just-right' way to address service classes and service plans from the instance spec.  Given the differences of opinion on that subject, I started thinking about supporting both of the following options:

(note, assumes we used the OSB ID of service classes/plans as the kubernetes name of these resources):

1.  Reference the kubernetes name (which in this case would be likely not to be human readable) of these resources
2.  Reference the human-readable name of the service class/plan at any given time (I think this is the mode @duglin seems to favor in [this comment](https://github.com/kubernetes-incubator/service-catalog/issues/802#issuecomment-317299121)
3.  Create 'alias' resources for service class and plan that have user-supplied, stable names that resolve to a specific OSB ID/kubernetes name of service class and plan

I'm not sure whether we would necessarily want to support *all* of these, but there are other variations of 'where does this service instance come from' (like user provided services, see #1010), that could also be modeled as siblings to these behaviors.

As I said [in this comment](https://github.com/kubernetes-incubator/service-catalog/issues/1010#issuecomment-318960677), my big takeaway is that we probably ought to use a union type to hold information about where a service instance comes from in order to add new options in a backward compatible way and clearly delineating exactly what option a user is requesting.

**Comment by MHBauer** [comment]
https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-architecture/rJGgMjrFIQc/rNm6uwlnBQAJ looks like the upstream discussion to follow.

**Comment by pmorie** [comment]
### Proposal

#### Assumptions

- name changes of services and plans are deprecated (as agreed to in OSB face to face) do not have to be accomodated in a first-class way
- users should never have to look at non-human-readable names of services or plans

#### Proposed Changes

- Move existing fields under `ServiceClass` and `ServicePlan` under new `Spec` structs and add `Status` to these objects
- Retain existing `ExternalID` fields and ensure these values are immutable for a given name of service and plan
- Kubernetes names:
  - `ServiceClass`: `$brokerName-$serviceClassName`
  - `ServicePlan`: `$brokerName-$serviceClassName-$servicePlanName`
- Add a new `Status.Inactive` (boolean) field to `ServiceClass` and `ServicePlan`
- If a broker changes the name of a service:
  - set the `Status.Inactive` field on that `ServiceClass`
  - set a status message saying that the service has been renamed to $new-name
  - do not allow new instances to be created of this service
- If a broker changes the name of a plan:
  - set the `Status.Inactive` field on that `ServicePlan`
  - set a status message saying that the plan has been renamed to $new-name
  - do not allow new instances to be created on this plan
- `ServiceInstance` keeps the same fields and uses the **full kubernetes names** to reference `ServiceClass` and `ServicePlan`


#### API Changes

```go
type ServiceClass struct {
	TypeMeta
	ObjectMeta

	Spec ServiceClassSpec
	Status ServiceClassStatus
}

type ServiceClassSpec struct {
	// all existing fields of ServiceClass here

}

type ServiceClassStatus struct {
	// Inactive indicates whether a service is inactive due to being deleted or renamed by the broker
	Inactive        boolean
	// InactiveMessage has information about why the ServiceClass was made inactive
	InactiveMessage *string
}

type ServicePlan struct {
	TypeMeta
	ObjectMeta

	Spec ServicePlanSpec
	Status ServicePlanStatus
}

type ServicePlanSpec struct {
	// all existing fields of ServicePlan here

}

type ServicePlanStatus struct {
	// Inactive indicates whether a plan is inactive due to being deleted or renamed by the broker
	Inactive        boolean
	// InactiveMessage has information about why the ServicePlan was made inactive
	InactiveMessage *string
}
```

#### Examples

Given a broker `test-broker` whose catalog contains a service `test-service` with plan `example-plan-1`, the following resources would be created:

```yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServiceClass
metadata:
  name: test-broker-test-service
spec:
  brokerName: test-broker
  externalID: d35b55b2-b1fd-4123-8045-5b9c619cb629
  description: "service description"
  bindable: true
---
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServicePlan
metadata:
  name: test-broker-test-service-example-plan-1
spec:
  brokerName: test-broker
  serviceClassName: test-broker-test-service
  externalID: 10e03cb7-b2cf-40dd-a954-16a382b92446
  description: "plan 1 description"
  free: true
```

A `ServiceInstance` referencing these would look like:

```yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServiceInstance
metadata:
  name: test-instance
  namespace: test-ns
spec:
  serviceClassName: test-broker-test-service
  planName: test-broker-test-service-example-plan-1
```

If the broker is **poorly behaved** and intentionally chooses to use **deprecated behaviors** and renames `example-plan-1` to `example-plan-2`, we would have the following `ServicePlan`s:

```yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServicePlan
metadata:
  name: test-broker-test-service-example-plan-1
spec:
  brokerName: test-broker
  serviceClassName: test-broker-test-service
  externalID: 10e03cb7-b2cf-40dd-a954-16a382b92446
  description: "plan 1 description"
  free: true
status:
  inactive: true
  inactiveMessage: "renamed to test-broker-test-service-example-plan-2"
---
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: ServicePlan
metadata:
  name: test-broker-test-service-example-plan-2
spec:
  brokerName: test-broker
  serviceClassName: test-broker-test-service
  externalID: 10e03cb7-b2cf-40dd-a954-16a382b92446
  description: "plan 1 description"
  free: true
```

**Comment by duglin** [comment]
Proposal discussed on today's call:  https://docs.google.com/presentation/d/1ZdRqLgAbE8uTQpJrXmdTucitPknURMoQITAJSxr3LhE/edit#slide=id.g25fb676d6e_0_91

**Comment by pmorie** [comment]
I might have another idea!

**Comment by pmorie** [comment]
I am starting to think that we should bail on solving all the complexities of name changes in the initial beta release and solve them later.  It's most important to get things into the hands of users and this seems to be a real bear of a problem to solve.  So, I'm not 100% sure, but I think we should consider doing it.

I do still have another idea I will write up ASAP, but I have exhausted my time budget for this problem for this week and need to move onto other things.

**Comment by pmorie** [comment]
Proposal:

- Use OSB ID for k8s names for now
- Add an `ExternalName` field to `ServiceClassSpec` and `ServicePlanSpec` that holds the value of the human-readable OSB name
- `ExternalName` should be a selectable field
- Users use the human-readable names to specify which service and plan they want via the `OSBServiceName` and `OSBPlanName` fields
- Add `ServiceClassRef` and `ServicePlanRef` fields into `ServiceInstanceSpec`
- Controller writes the `ServiceClassRef` and `ServicePlanRef` fields if they are not set
- Controller updates the `ServicePlanRef` field if the `OSBPlanName` is updated by the user

```go
type ServiceInstanceSpec struct {
	// OSBServiceName is the human-readable name of the service
	// as reported by the broker.
	OSBServiceName string
	// OSBPlanName is the human-readable name of the plan
	// as reported by the broker.
	OSBPlanName string

	// ServiceClassRef is a reference to the ServiceClass
	// that the user selected.
	ServiceClassRef *ObjectReference
	// ServicePlanRef is a reference to the ServicePlan
	// that the user selected.
	ServicePlanRef  *ObjectReference
}

type ServiceClassSpec struct {
	ExternalID   string
	ExternalName string
}

type ServicePlanSpec struct {
	ExternalID   string
	ExternalName string
}
```

**Comment by vaikas-google** [comment]
@pmorie @duglin 
Do we still need ExternalID fields in Service[Class|Plan]Spec? Seems like that's simple .Name in both cases if we use the ID as the name and it's immutable?

**Comment by pmorie** [comment]
@vaikas-google I do want both external ID and Name fields so we can support different naming strategies in the future

**Comment by vaikas-google** [comment]
ack
